lib/dialects/abstract/query-generator.js

```diff
@@ -1387,11 +1387,20 @@ module.exports = (function() {
       }
 
       if (options.lock && this._dialect.supports.lock) {
-        if (options.lock === 'SHARE') {
+        var lock = options.lock;
+        if (typeof options.lock === 'object') {
+          lock = options.lock.level;
+        }
+        if (this._dialect.supports.lockKey && (lock === 'KEY SHARE' || lock === 'NO KEY UPDATE')) {
+          query += ' FOR ' + lock;
+        } else if (lock === 'SHARE') {
           query += ' ' + this._dialect.supports.forShare;
         } else {
           query += ' FOR UPDATE';
         }
+        if (this._dialect.supports.lockOf && options.lock.of instanceof Model) {
+          query += ' OF ' + this.quoteTable(options.lock.of.name);
```

Comment:
`optinos.lock.of.tableName`, name might be different from tableName.
And what about schema support? Should it actually be `options.lock.of.getTableName()`?